src/libraries/System.Text.Json/src/System/Text/Json/Serialization/DefaultReferenceResolver.cs

```diff
@@ -3,80 +3,62 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using System.Diagnostics;
 
 namespace System.Text.Json.Serialization
 {
     /// <summary>
     /// The default ReferenceResolver implementation to handle duplicate object references.
     /// </summary>
-    /// <remarks>
-    /// It is currently a struct to save one unnecessary allcation while (de)serializing.
-    /// If we choose to expose the ReferenceResolver in a future, we may need to create an abstract class/interface and change this type to become a class that inherits from that abstract class/interface.
-    /// </remarks>
-    internal struct DefaultReferenceResolver
+    internal class DefaultReferenceResolver : ReferenceResolver
     {
         private uint _referenceCount;
         private readonly Dictionary<string, object>? _referenceIdToObjectMap;
         private readonly Dictionary<object, string>? _objectToReferenceIdMap;
 
-        public DefaultReferenceResolver(bool writing)
+        public DefaultReferenceResolver()
         {
             _referenceCount = default;
 
-            if (writing)
-            {
-                // Comparer used here to always do a Reference Equality comparison on serialization which is where we use the objects as the TKey in our dictionary.
-                _objectToReferenceIdMap = new Dictionary<object, string>(ReferenceEqualityComparer.Instance);
-                _referenceIdToObjectMap = null;
-            }
-            else
-            {
-                _referenceIdToObjectMap = new Dictionary<string, object>();
-                _objectToReferenceIdMap = null;
-            }
+            // Comparer used here to always do a Reference Equality comparison on serialization which is where we use the objects as the TKey in our dictionary.
+            _objectToReferenceIdMap = new Dictionary<object, string>(ReferenceEqualityComparer.Instance);
+            _referenceIdToObjectMap = new Dictionary<string, object>();
```

Comment:
Because there is no way to tell when to use one or another anymore. 

Previous to these changes, the constructor had a bool that indicated whether this was serialization or deserialization but that seemed like implementation detail leaking out to the API and I decided to cut that out. Maybe we could add an overload that takes the bool back if instantiating both dictionaries hurts the performance too much.

Reply:
As we discussed offline, I added the internal method `CreateResolvr(bool writing)` to optimize for `ReferenceHandler.Preserve`, now it will be able to know when to instantiate one dictionary or the other and avoid always instantiating twice.